### PR TITLE
Gracefully handle syntax error

### PR DESCRIPTION
**Pull request by kaisecheng**
Fixed: #32 

Catch syntax error in register phase to let other healthy pipelines continue

### CODE REVIEW

File: spec/filters/ruby_spec.rb [hunk 1]
```diff
@@ -56,7 +56,7 @@
         # non-pretty json does not contain \n
         insist { subject.get("pretty").count("\n") } == 0
         # usage of JSON.parse here is to avoid parser-specific order assertions
-        insist { JSON.parse(subject.get("pretty")) } == JSON.parse("{\"message\":\"hello world\",\"mydate\":\"2014-09-23T00:00:00-0800\",\"@version\":\"1\",\"@timestamp\":\"2014-09-23T08:00:00.000Z\"}")
+        insist { JSON.parse(subject.get("pretty")) } == JSON.parse("{\"message\":\"hello world\",\"mydate\":\"2014-09-23T00:00:00-0800\",\"@version\":\"1\",\"@timestamp\":\"#{get_logstash_timestamp("2014-09-23T08:00:00.000Z")}\"}")
```
Comment (kaisecheng) [inline spec/filters/ruby_spec.rb:59]: fix the test with Logstash 8 nanosecond-precision timetamps
https://github.com/elastic/logstash/pull/12797

File: lib/logstash/filters/ruby.rb [hunk 2]
```diff
@@ -66,6 +66,11 @@ def register
     else
       @logger.fatal("You must either use an inline script with the \"code\" option or a script file using \"path\".")
     end
+  rescue SyntaxError => e
+    @logger.error("Syntax Error in register #{e.message}",
+                  :class     => e.class.name,
+                  :backtrace => e.backtrace)
+    fail("Ruby script syntax error")
```
Comment (kares) [inline lib/logstash/filters/ruby.rb:73]: might be better to handle `ScriptError` in general (covers `ScriptError` as well as `LoadError` etc.) e.g.
```suggestion
  rescue ScriptError => e
    @logger.error("error in register", message: e.message, exception: e.class, backtrace: e.backtrace)
    fail("unexpected error: #{e}")
```
... just a suggestion I did not test this - might require more refining

Comment (kaisecheng) [reply lib/logstash/filters/ruby.rb]: `ScriptError` is a `StandardError`. So, rescue ScriptError and SyntaxError

**Comment by andsel** [comment]
I've built the gem locally (the version `3.1.8`) and installed in a fresh  Logstash `7.16.1`.
- configured `config/pipelines.yml` with:
```
- pipeline.id: broken_ppln
  path.config: "/tmp/fail_ruby_filter_pipeline.conf"

- pipeline.id: success_ppln
  config.string: "input { generator {} } filter { sleep { time => 1 } } output { stdout { codec => dots } }"
```
- the content of `/tmp/fail_ruby_filter_pipeline.conf` is:
```
input { stdin {} }
filter {
  ruby { code => "[].size (1,23)" }
}
output { stdout { codec => rubydebug } }
```
- run Logstash with `bin/logstash`

What I would have expected was to see the `success_ppln` producing dots, but I got a crash of Logstash:
```sh
andrea@kalispera logstash-7.16.1 $ bin/logstash
Using JAVA_HOME defined java: /home/andrea/.sdkman/candidates/java/current
WARNING: Using JAVA_HOME while Logstash distribution comes with a bundled JDK.
DEPRECATION: The use of JAVA_HOME is now deprecated and will be removed starting from 8.0. Please configure LS_JAVA_HOME instead.
OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
Sending Logstash logs to /home/andrea/workspace/elastic_products/logstash-7.16.1/logs which is now configured via log4j2.properties
[2022-01-24T16:31:26,519][INFO ][logstash.runner          ] Log4j configuration path used is: /home/andrea/workspace/elastic_products/logstash-7.16.1/config/log4j2.properties
[2022-01-24T16:31:26,530][INFO ][logstash.runner          ] Starting Logstash {"logstash.version"=>"7.16.1", "jruby.version"=>"jruby 9.2.20.1 (2.5.8) 2021-11-30 2a2962fbd1 OpenJDK 64-Bit Server VM 11.0.11+9 on 11.0.11+9 +indy +jit [linux-x86_64]"}
[2022-01-24T16:31:27,946][INFO ][logstash.agent           ] Successfully started Logstash API endpoint {:port=>9600, :ssl_enabled=>false}
[2022-01-24T16:31:28,771][INFO ][org.reflections.Reflections] Reflections took 85 ms to scan 1 urls, producing 119 keys and 417 values 
[2022-01-24T16:31:30,121][INFO ][logstash.javapipeline    ][success_ppln] Starting pipeline {:pipeline_id=>"success_ppln", "pipeline.workers"=>12, "pipeline.batch.size"=>125, "pipeline.batch.delay"=>50, "pipeline.max_inflight"=>1500, "pipeline.sources"=>["config string"], :thread=>"#<Thread:0x7d588d67 run>"}
[2022-01-24T16:31:30,127][INFO ][logstash.javapipeline    ][broken_ppln] Pipeline terminated {"pipeline.id"=>"broken_ppln"}
warning: thread "[broken_ppln]-pipeline-manager" terminated with exception (report_on_exception is true):
SyntaxError: (ruby filter code):2: syntax error, unexpected ','
 [].size (1,23) 
           ^
                     eval at org/jruby/RubyKernel.java:1048
                 register at /home/andrea/workspace/elastic_products/logstash-7.16.1/vendor/local_gems/e3b5beb4/logstash-filter-ruby-3.1.8/lib/logstash/filters/ruby.rb:63
                 register at org/logstash/config/ir/compiler/AbstractFilterDelegatorExt.java:75
         register_plugins at /home/andrea/workspace/elastic_products/logstash-7.16.1/logstash-core/lib/logstash/java_pipeline.rb:232
                     each at org/jruby/RubyArray.java:1821
         register_plugins at /home/andrea/workspace/elastic_products/logstash-7.16.1/logstash-core/lib/logstash/java_pipeline.rb:231
  maybe_setup_out_plugins at /home/andrea/workspace/elastic_products/logstash-7.16.1/logstash-core/lib/logstash/java_pipeline.rb:590
            start_workers at /home/andrea/workspace/elastic_products/logstash-7.16.1/logstash-core/lib/logstash/java_pipeline.rb:244
                      run at /home/andrea/workspace/elastic_products/logstash-7.16.1/logstash-core/lib/logstash/java_pipeline.rb:189
                    start at /home/andrea/workspace/elastic_products/logstash-7.16.1/logstash-core/lib/logstash/java_pipeline.rb:141
[2022-01-24T16:31:30,158][ERROR][logstash.agent           ] Failed to execute action {:id=>:broken_ppln, :action_type=>LogStash::ConvergeResult::FailedAction, :message=>"Could not execute action: PipelineAction::Create<broken_ppln>, action_result: false", :backtrace=>nil}
[2022-01-24T16:31:30,326][FATAL][org.logstash.Logstash    ] Logstash stopped processing because of an error: (SyntaxError) (ruby filter code):2: syntax error, unexpected ','
 [].size (1,23) 
           ^
org.jruby.exceptions.SyntaxError: (SyntaxError) (ruby filter code):2: syntax error, unexpected ','
 [].size (1,23) 
           ^
	at org.jruby.RubyKernel.eval(org/jruby/RubyKernel.java:1048) ~[jruby-complete-9.2.20.1.jar:?]
	at home.andrea.workspace.elastic_products.logstash_minus_7_dot_16_dot_1.vendor.local_gems.e3b5beb4.logstash_minus_filter_minus_ruby_minus_3_dot_1_dot_8.lib.logstash.filters.ruby.register(/home/andrea/workspace/elastic_products/logstash-7.16.1/vendor/local_gems/e3b5beb4/logstash-filter-ruby-3.1.8/lib/logstash/filters/ruby.rb:63) ~[?:?]
	at org.logstash.config.ir.compiler.AbstractFilterDelegatorExt.register(org/logstash/config/ir/compiler/AbstractFilterDelegatorExt.java:75) ~[logstash-core.jar:?]
	at home.andrea.workspace.elastic_products.logstash_minus_7_dot_16_dot_1.logstash_minus_core.lib.logstash.java_pipeline.register_plugins(/home/andrea/workspace/elastic_products/logstash-7.16.1/logstash-core/lib/logstash/java_pipeline.rb:232) ~[?:?]
	at org.jruby.RubyArray.each(org/jruby/RubyArray.java:1821) ~[jruby-complete-9.2.20.1.jar:?]
	at home.andrea.workspace.elastic_products.logstash_minus_7_dot_16_dot_1.logstash_minus_core.lib.logstash.java_pipeline.register_plugins(/home/andrea/workspace/elastic_products/logstash-7.16.1/logstash-core/lib/logstash/java_pipeline.rb:231) ~[?:?]
	at home.andrea.workspace.elastic_products.logstash_minus_7_dot_16_dot_1.logstash_minus_core.lib.logstash.java_pipeline.maybe_setup_out_plugins(/home/andrea/workspace/elastic_products/logstash-7.16.1/logstash-core/lib/logstash/java_pipeline.rb:590) ~[?:?]
	at home.andrea.workspace.elastic_products.logstash_minus_7_dot_16_dot_1.logstash_minus_core.lib.logstash.java_pipeline.start_workers(/home/andrea/workspace/elastic_products/logstash-7.16.1/logstash-core/lib/logstash/java_pipeline.rb:244) ~[?:?]
	at home.andrea.workspace.elastic_products.logstash_minus_7_dot_16_dot_1.logstash_minus_core.lib.logstash.java_pipeline.run(/home/andrea/workspace/elastic_products/logstash-7.16.1/logstash-core/lib/logstash/java_pipeline.rb:189) ~[?:?]
	at home.andrea.workspace.elastic_products.logstash_minus_7_dot_16_dot_1.logstash_minus_core.lib.logstash.java_pipeline.start(/home/andrea/workspace/elastic_products/logstash-7.16.1/logstash-core/lib/logstash/java_pipeline.rb:141) ~[?:?]
andrea@kalispera logstash-7.16.1 $ 
```

**Comment by andsel** [comment]
Mmm I used a version before f05199075e891a22a5faad1aef2c4ac685e8ee78

**Comment by kares** [comment]
> Mmm I used a version before f051990

Interesting ...
```
>> SyntaxError.superclass
=> ScriptError
>> ScriptError.superclass
=> Exception
```
... think there's a `ScriptError` within the plugin - what we really want is to catch `::ScriptError` than

File: lib/logstash/filters/ruby.rb [hunk 3]
```diff
@@ -66,6 +66,9 @@ def register
     else
       @logger.fatal("You must either use an inline script with the \"code\" option or a script file using \"path\".")
     end
+  rescue ScriptError, SyntaxError => e
```
Comment (kares) [inline lib/logstash/filters/ruby.rb:69]: sorry I did not realize there's a `ScriptError` in scope, what I really wanted is:
```suggestion
  rescue ::ScriptError => e # SyntaxError, LoadError etc.
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,6 @@
-## x.y.z
+## 3.1.8
  - [DOC] Added doc to describe the option `tag_with_exception_message`[#62](https://github.com/logstash-plugins/logstash-filter-ruby/pull/62)
+ - Fix SyntaxError handling so other pipelines can shut down gracefully
```
[the remark this answers is not in the document]
Comment (kaisecheng) [reply CHANGELOG.md:3]: oh! good catch!

**Review by kares — APPROVED** [review]
:bow_and_arrow:

**Pull request merged**